lxd/instances_post.go

```diff
@@ -880,65 +872,17 @@ func instancesPost(d *Daemon, r *http.Request) response.Response {
 			return err
 		}
 
-		var allMembers []db.NodeInfo
-
-		if clustered && !clusterNotification {
-			clusterGroupsAllowed = shared.SplitNTrimSpace(targetProject.Config["restricted.cluster.groups"], ",", -1, true)
-
-			// Check manual cluster member targeting restrictions.
-			err = project.CheckClusterTargetRestriction(r, targetProject, target)
-			if err != nil {
-				return err
-			}
-
-			allMembers, err = tx.GetNodes(ctx)
-			if err != nil {
-				return fmt.Errorf("Failed getting cluster members: %w", err)
-			}
-
-			if targetMember != "" {
-				// Find target member.
-				for i := range allMembers {
-					if allMembers[i].Name == targetMember {
-						targetMemberInfo = &allMembers[i]
-						break
-					}
-				}
-
-				if targetMemberInfo == nil {
-					return api.StatusErrorf(http.StatusNotFound, "Cluster member not found")
-				}
-
-				// If restricted groups are specified then check member is in at least one of them.
-				if shared.IsTrue(targetProject.Config["restricted"]) && len(clusterGroupsAllowed) > 0 {
-					found := false
-					for _, memberGroupName := range targetMemberInfo.Groups {
-						if shared.StringInSlice(memberGroupName, clusterGroupsAllowed) {
-							found = true
-							break
-						}
-					}
-
-					if !found {
-						return api.StatusErrorf(http.StatusForbidden, "Project isn't allowed to use this cluster member")
-					}
-				}
-			} else if targetGroup != "" {
-				// If restricted groups are specified then check the requested group is in the list.
-				if shared.IsTrue(targetProject.Config["restricted"]) && len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(targetGroup, clusterGroupsAllowed) {
-					return api.StatusErrorf(http.StatusForbidden, "Project isn't allowed to use this cluster group")
-				}
+		allMembers, err := tx.GetNodes(ctx)
```

Comment:
I think we've lost some optimisations here, because previously these queries for tx.GetNodes and the ones in project.CheckTarget were done only when `if clustered && !clusterNotification {`.

Now, from my reading of this, it seems we are doing them both on the initial request *and* on the subsequent forwarded request when it arrives at the target member. This seems unnecessary to me. 

Was there a reason you needed to change this?

Reply:
When comparing with master, looks like the check for `clustered && !clusterNotification` for `GetNodes` got lost in between refactoring and moving the the checks for the target. Below it was covered by the `clustered && !clusterNotification && targetMemberInfo == nil` check we discussed earlier.

To be as close as possible to master, for the fetching of the nodes and target checks I'll add the cluster and notification check.

Reply:
I'll resolve this since it is not part of the PR anymore.